pkg/gadgets/top/ebpf/tracer/tracer.go

```diff
@@ -0,0 +1,249 @@
+
+	return t, nil
+}
+
+func (t *Tracer) start() error {
+	// Enable stats collection
+	err := bpfstats.EnableBPFStats()
+	if err != nil {
+		return err
+	}
+
+	t.useFallbackIterator = false
+
+	// To resolve pids, we will first try to iterate using a bpf
+	// program. If that doesn't work, we will fall back to scanning
+	// all used fds in all processes /proc/$pid/fdinfo/$fd.
+	iter, err := piditer.NewTracer()
+	if err != nil {
+		t.useFallbackIterator = true
+	} else {
+		t.iter = iter
+	}
+
+	return nil
+}
+
+func (t *Tracer) Stop() {
+	close(t.done)
+
+	if t.iter != nil {
+		t.iter.Close()
+	}
+
+	bpfstats.DisableBPFStats()
+}
+
+func getPidMapFromPids(pids []*piditer.PidIterEntry) map[uint32][]*types.PidInfo {
+	pidmap := make(map[uint32][]*types.PidInfo)
+	for _, e := range pids {
+		if _, ok := pidmap[e.ProgID]; !ok {
+			pidmap[e.ProgID] = make([]*types.PidInfo, 0, 1)
+		}
+		pidmap[e.ProgID] = append(pidmap[e.ProgID], &types.PidInfo{
+			Pid:  e.Pid,
+			Comm: e.Comm,
+		})
+	}
+	return pidmap
+}
+
+func getPidMapFromProcFs() map[uint32][]*types.PidInfo {
+	pidmap := make(map[uint32][]*types.PidInfo)
+	processes, err := os.ReadDir("/proc/")
+	if err == nil {
+		for _, p := range processes {
+			if !p.IsDir() {
+				continue
+			}
+			_, err := strconv.Atoi(p.Name())
+			if err != nil {
+				continue
+			}
+			fdescs, err := os.ReadDir(filepath.Join("/proc", p.Name(), "fdinfo"))
+			if err != nil {
+				continue
+			}
+			for _, fd := range fdescs {
+				f, err := os.Open(filepath.Join("/proc", p.Name(), "fdinfo", fd.Name()))
+				if err != nil {
+					continue
+				}
+				sc := bufio.NewScanner(f)
+				for sc.Scan() {
+					if strings.HasPrefix(sc.Text(), "prog_id:") {
+						pid, _ := strconv.ParseUint(p.Name(), 10, 32)
+						progID, _ := strconv.ParseUint(strings.TrimSpace(strings.Split(sc.Text(), ":")[1]), 10, 32)
+						if _, ok := pidmap[uint32(progID)]; !ok {
+							pidmap[uint32(progID)] = make([]*types.PidInfo, 0, 1)
+						}
+						comm, _ := os.ReadFile(filepath.Join("/proc", p.Name(), "comm"))
+						pidmap[uint32(progID)] = append(pidmap[uint32(progID)], &types.PidInfo{
+							Pid:  uint32(pid),
+							Comm: strings.TrimSpace(string(comm)),
+						})
+						break
+					}
+				}
+				f.Close()
+			}
+		}
+	}
+	return pidmap
+}
+
+func (t *Tracer) nextStats() ([]types.Stats, error) {
+	stats := make([]types.Stats, 0)
+
+	var err error
+	var prog *ebpf.Program
+	var pids []*piditer.PidIterEntry
+	curID := ebpf.ProgramID(0)
+	nextID := ebpf.ProgramID(0)
+
+	for {
+		nextID, err = ebpf.ProgramGetNextID(curID)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("could not get next program ID: %w", err)
+		}
+		if nextID <= curID {
+			break
+		}
+		curID = nextID
+		prog, err = ebpf.NewProgramFromID(curID)
+		if err != nil {
+			continue
+		}
+		pi, err := prog.Info()
+		if err != nil {
+			prog.Close()
+			continue
+		}
+		rt, _ := pi.Runtime()
+		rc, _ := pi.RunCount()
```

Comment:
It seems this logic is wrong here as it'll get the numbers from the time `EnableBPFStats()` instead of the last interval. I think it's needed to keep a map with the id of the program and the values when the gadget started to be able to calculate the difference.

Reply:
Hmm. Right now the runtime is used the same way that the system command `top` uses it - it shows the total cpu time - (but yes, in our case only since collecting the stats were enabled/when they were enabled).

But yeah, then, there's a difference: by default, the system command `top` sorts by current cpu usage and not the total cpu time. Current usage is a value that we currently don't have - and for that, as you said, we would need to store the old value in a map and compare. We could then show the time used in the current interval or - most likely taking into account the number of CPU cores - calculate the cpu percentage as well.

I'll look into it, if we want to have both values - but I wouldn't scratch the total runtime from the stats, as I think that it's more helpful to have the absolute numbers when you look at longer timespans.

Reply:
I left the values for the total numbers in the event/stats, but removed them from the default output. Instead it now uses the "last interval" values for everything by default.